server/aesdsocket.c

```diff
@@ -80,12 +84,141 @@ void signal_init()
+    int prev_size = BUF_SIZE, total_len = 0;
+    int bound = FALSE;         //flag for bind condition
+    char ip4[INET_ADDRSTRLEN]; // space to hold the IPv4 string
+
+    memset(&recv_buf, '\0', BUF_SIZE);
+
+    while (!recv_complete)
+    {
+        ssize_t recv_bytes;
+        recv_len = 0;
+
+        if ((recv_bytes = recv(node->cfd, &recv_buf, BUF_SIZE, 0)) == -1)
+        {
+            syslog(LOG_ERR, "recv\n");
+            exit(EXIT_FAILURE);
+        }
+        for (i = 0; i < recv_bytes; i++)
+        {
+            recv_len++;
+
+            if (recv_buf[i] == '\n')
+            {
+                recv_complete = TRUE;
+                break;
+            }
+        }
+
+        if (first_cycle)
+        {
+            tx_buf = (char *)malloc(recv_len + 1);
+
+            if (tx_buf == NULL)
+            {
+                syslog(LOG_ERR, "malloc");
+                exit(EXIT_FAILURE);
+            }
+
+            memset(tx_buf, '\0', recv_len);
+
+            first_cycle = FALSE;
+
+            total_len = recv_len;
+        }
+        else
+        {
+            // printf("Prev size: %d\n", prev_size);
+            if ((temp_ptr = realloc(tx_buf, prev_size + recv_len + 1)) == NULL)
+            {
+                syslog(LOG_ERR, "realloc");
+                exit(EXIT_FAILURE);
+            }
+            else
+            {
+                tx_buf = temp_ptr;
+                memset(tx_buf + prev_size, '\0', recv_len);
+                prev_size += recv_len;
+            }
+            total_len = prev_size;
+        }
+
+        strncat(tx_buf, recv_buf, recv_len); //copy present contents
+    }
+
+    //Locking before accessing file
+    pthread_mutex_lock(&mutex);
+
+    fd = open(MY_SOCK_PATH, O_CREAT | O_RDWR | O_APPEND, 0644);
+    if (fd == -1)
+    {
+        syslog(LOG_ERR, "open");
+        exit(EXIT_FAILURE);
+    }
+
+    syslog(LOG_DEBUG, "Opened file\n");
+
+    //Appending Packet to the file
+    lseek(fd, 0, SEEK_END);
+    int bytes = write(fd, tx_buf, total_len);
+    if (bytes == -1)
+    {
+        syslog(LOG_ERR, "write\n");
+        exit(EXIT_FAILURE);
+    }
+    // printf("Writing %d bytes\n", bytes);
+
+    lseek(fd, 0, SEEK_SET);
+
+    while ((bytes = read(fd, recv_buf, BUF_SIZE)) != 0)
+    {
+        //bytes = read(fd, recv_buf, BUF_SIZE);
+        if (bytes == -1)
+        {
+            syslog(LOG_ERR, "read\n");
+            exit(EXIT_FAILURE);
+        }
+        //printf("Reading %d bytes\n", bytes);
+        if (send(node->cfd, &recv_buf, bytes, 0) == -1)
+        {
+            syslog(LOG_ERR, "send\n");
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    free(tx_buf);
+    close(node->cfd);
+    node->cfd = -1; //reset to indicate fd already clsoed (signal handler)
+    syslog(LOG_DEBUG, "Closed connection from %s\n", ip4);
+
+    //Unlock mutex after completing transactions with file and updating thread status
+    pthread_mutex_unlock(&mutex);
```

Comment:
@kamini2801  Aren't you holding mutex for too long? Is this locking code? or locking data?

Reply:
@Maitreyee2095 post read all three operations happen on local arguments (tx_buff, node pointer) so the mutex can be unlocked right after the while read. Will make that change during the next implementation, thanks